rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,504 @@
+
+This RFC will also formally define some GraphQL APIs as some of them were loosely defined in the past. This would give us more robustness for this RFC's design and less reason to refactor the architecture in the long run.
+
+# Detailed Design
+
+The first section will talk about the requirements and basic primitives that a client should be able to do when interacting with the protocol. This will make a good segway for explaining the proposed design. The last section will discuss consistency issues that could occur with this design.
+
+<a href="requirements"></a>
+
+## Requirements
+
+__NOTE: The types presented in OCaml code are disk representations of the types__
+
+In the Coda blockchain, there would be many different transactions sent and received from various people in the network. We would only be interested in hearing about several transactions involving certain people (i.e. friends). Therefore, a client should only keep a persistent record of transactions involving a white list of people and they should be able to look up the contents of these transactions from a container. Here are the records for a transaction. Note that transactions could be either `user_commands` or `fee_transfers` for this design:
+
+```ocaml
+module User_commands = struct
+  type t = {
+    id: string;
+    is_delegation : bool;
+    nonce: Unsigned64.t;
+    from: Public_key.Compressed.t;
+    to: Public_key.Compressed.t;
+    amount: Unsigned64.t;
+    fee: Unsigned64.t;
+    memo: string
+  }
+end
+
+module Fee_transfer {
+  id: string;
+  fee: Int64.t;
+  receiver: Public_key.Compressed.t;
+}
+```
+
+Additionally, clients should be able to know which `external_transition`s contains a transaction. These transitions can give us useful information about a transaction, like the number of block confirmations that is on top of a certain transaction and the "status" of a transaction. Therefore, a client should be able to store a record of these interested external_transitions and look them up easily in a container. Here is the type of an external_transition:
+
+```ocaml
+module Protocol_state = struct
+  type t = {
+    previousStateHash: string;
+    blockchainState: BlockchainState;
+    length: Int64!
+  }
+end
+
+
+module External_transition = struct
+  state_hash: string;
+  creator: Public_key.Compressed.t
+  protocol_state: Protocol_state.t [not null]
+  is_snarked bool [not null]
```

Comment:
FWIW, you could attempt to quantify transactions into "proof batches" based on how they are added to the staged ledger diff, and keep track of the current height of "completed proof batches" based on proofs that have been included into the blockchain proof via an emitted ledger proof. That would help you to easily determine whether a transaction has been proven or not in a simple O(1) queryable way.

Reply:
Yes, I meant snarked transactions (i.e. the transactions are snarked when a ledger proof is emitted). Sorry for the confusion.

Reply:
In that case, you will need to change this structure to something like what I described above since the transactions in a transition are not necessarily snarked in the same proof.